### ISSUE TITLE
How to implement unit conversion?

### ISSUE THREAD

**Issue by mhvk** [astropy/quantity-2.0#4]
A question is how to best do unit conversion. We could follow astropy `Quantity` have methods like `.to()` and perhaps `.to_value()`, or could let the `Quantity` class not worry about it. E.g., astropy's `Unit` allows construction with `array * unit` (array copied) or `array << unit` (array not copied). Currently, `Quantity2.0` does not copy, so the two might become equivalent.

Following astropy, one could imagine doing unit conversion via `q << new_unit` (which could be implemented either via `Quantity.__lshift__` or `Unit.__rlshift__`). 

As an operator equivalent of `.to_value()`, one option might be to use the reverse direction, `q >> value_unit` (shift out the value with the given unit, where `<<` shifts the input to the unit).

Another might be `q // unit` - the logic here would be that one "truncate after dividing by the unit", inspired by `q1 // q2` being the truncated dimensionless ratio of the two quantities [1]. Again, it could either be `Quantity.__floordiv__` or `Unit.__rfloordiv__` that takes care of this. A disadvantage is that `q1 // unit` is then different from `q1 // (1*unit)`, but perhaps not too bad -- we may want to avoid more generally to have units behave like quantities. Also, it might then make sense to have `q1 % unit` return `0*q1.unit` so that `q1 % unit + (q1 // unit) * unit== q` will always hold. (If we go this route, it might make sense to let `q1 // q2` return a plain array rather than a dimensionless `Quantity`.)

Alternatively, as apparently is done by Julia, we could just leave it to the unit to have some `convert` function (a bit like the array API has `astype` for conversion of data types).

1. In fairness, I should point out that I suggested `// unit` a long time ago for astropy (https://github.com/astropy/astropy/pull/1412 and it was roundly rejected.

**Comment by byrdie** [comment]
Based on my troubles in astropy/#16515, I was left wishing that unit conversion could be done by a ``convert()`` function which would allow for multiple dispatch. This would make solving astropy/#16515 trivial for `named-arrays` and other duck types which would like to wrap Quantity-2.0. 

I still think that we should have a `to()` method and `<<` operator support, but that they should just be syntactic sugar for the `convert()` function.

**Comment by mhvk** [comment]
@byrdie - does your `convert` version possibly operate on quantity duck types, or just on pure arrays? I can see the various unit packages agreeing on having a `u.convert(from_unit, to_unit, value)`. But perhaps you are thinking more of `u.convert(to_unit, quantity)`, where the overloading allows one to specific where the unit and value of the given quantity should be gotten from?

**Comment by nstarman** [comment]
I suspect using a non built-in implementation of multiple dispatch isn't going to happen. But single-dispatch totally could. We can dispatch on the quantity type, not the unit type.
If you want to see an example Python library where `uconvert` is the primary function and the methods on Quantity are syntactic sugar, check out `unxt`, which I built to make Astropy units work with JAX.

**Comment by byrdie** [comment]
@mhvk and @nstarman, I suppose I haven't thought through the signature very carefully, but my knee-jerk reaction was `u.convert(quantity, to_unit, equivalencies=...)`, where we allowed multiple dispatch on both `quantity` and the elements of `equivalencies` through machinery similar to __array_function__. Obviously this isn't going to work out of the box since `equivalencies` isn't an array type, but maybe there's a way to tweak it or inspect it?

Apologies if I'm asking for too many changes, it's just that this is the only part of the astropy Quantity API that I have not been able to override with `named-arrays`, and I'm rather frustrated by it.

**Comment by byrdie** [comment]
@nstarman, thanks for pointing out `unxt`, I've been following that project for a while and I'm very interested in trying to wrap it with my `named-arrays` library but I haven't had a chance to try it yet.

**Comment by mhvk** [comment]
p.s. More relevant for `Quantity 2.0`, perhaps: if your `named-arrays` implements the Array API, then in principle `Quantity 2.0` should "just work" with it. I think as is some tests might need changes since they create arrays without worrying about names, but I think this is do-able. Feel free to have a look!

**Comment by nstarman** [comment]
> E.g., I don't quite understand why equivalencies would need to be overwritten if one has .to(), .to_value(), and .value defined just like astropy's Quantity.

Agreed. The equivalencies should just be a function that's operated on the Quantity in a array-type agnostic way.

**Comment by byrdie** [comment]
In my opinion, it would be nice to give the opportunity for downstream classes to override the unit conversion behavior. In the `named-arrays` / `xarray` case, this would be used to align the axes between `value` and `equivalencies` before applying the conversion, but I understand that this is a special case that you might not want to address right now. I'll continue this discussion in astropy/astropy#16518

**Comment by mhvk** [comment]
I think if we write equivalencies sufficiently carefully, they should internally have values (not quantities) that can be `named-array`. But let's indeed discuss over at astropy (in a *new* issue, not a closed PR!)
